### PR TITLE
Event loop refactor

### PR DESCRIPTION
**Pull request by TornaxO7**
# Description

Just a little refactoring to the event loop.

Fixes #(issue) None

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Updated user documentation:

None

# Checklist:

- [x] Ran `make test-full` locally with no errors or warnings reported
- [x] Manual page has been updated accordingly
- [x] Wiki pages have been updated accordingly (to perform **after** merge)

### CODE REVIEW

File: leftwm-core/src/event_loop.rs [hunk 1]
```diff
@@ -1,123 +1,194 @@
...
-        let mut event_buffer = vec![];
-        loop {
-            if self.state.mode == Mode::Normal {
-                state_socket.write_manager_state(&self.state).await.ok();
-            }
-            self.display_server.flush();
+        let mut event_buffer: Vec<DisplayEvent> = vec![];
+        while self.should_keep_running(&mut state_socket).await {
+            self.update_manager_state(&mut state_socket).await;
+            self.display_server.flush(); // is that needed?
```
Comment (AethanFoot) [inline leftwm-core/src/event_loop.rs:39]: Seems to be needed (i found with lefthk that it seemed to make a difference, atleast on startup).

Comment (TornaxO7) [reply leftwm-core/src/event_loop.rs:39]: > i found with lefthk that it seemed to make a difference, atleast on startup

may I ask what kind of difference it does?

Comment (AethanFoot) [reply leftwm-core/src/event_loop.rs:39]: Xlib events not coming in, i suppose sometimes they may get blocked up and needs flushing through to get them going again

Comment (TornaxO7) [reply leftwm-core/src/event_loop.rs]: alright, I removed the comment

File: rust-toolchain.toml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+[toolchain]
+channel = "stable"
```
Comment (VuiMuich) [inline rust-toolchain.toml:2]: Sneaking this in again through the backdoor huh?
:rofl: jk, keep it in.

Comment (TornaxO7) [reply rust-toolchain.toml:2]: oh bruh, I forgot it. Sorry! :rofl: 

> rofl jk, keep it in.

oh lol, if you insist, I'd like to :)

**Comment by TornaxO7** [comment]
May I ask what it's not merged yet?

**Comment by VuiMuich** [comment]
We are about to do. I just wanted to give some others a chance to review as well. Sorry for taking this long.

As you may have noticed, we are more on the "chaotic good" side of the dnd spectrum of repos :grimacing: so that's why we did not assign some review tasks etc.

**Comment by TornaxO7** [comment]
>  I just wanted to give some others a chance to review as well. Sorry for taking this long.

ah, understandable :+1:

File: leftwm-core/src/utils/state_socket.rs [hunk 3]
```diff
@@ -61,7 +61,9 @@ impl StateSocket {
             let mut json = serde_json::to_string(&state)?;
             json.push('\n');
             let mut state = self.state.lock().await;
-            if json != state.last_state {
+
+            let state_changed = json != state.last_state;
+            if state_changed {
```
Comment (hertg) [inline leftwm-core/src/utils/state_socket.rs:66]: I think that's good for code readability. My guess is the compiler just optimizes that variable away anyway right?

File: leftwm/src/bin/leftwm-worker.rs [hunk 4]
```diff
@@ -3,23 +3,22 @@ use slog::{o, Drain};
 use std::panic;
 
 fn main() {
-    //let _log_guard = setup_logfile();
-    let _log_guard = setup_logging();
+    let _log_guard = setup_logfile();
+    // let _log_guard = setup_logging();
```
[the remark this answers is not in the document]
Comment (TornaxO7) [reply leftwm/src/bin/leftwm-worker.rs:7]: Yes I did!

File: leftwm-core/src/utils/state_socket.rs (hunk 3 above)
Comment (TornaxO7) [reply leftwm-core/src/utils/state_socket.rs:66]: > My guess is the compiler just optimizes that variable away anyway right?

yes it does!

**Comment by AethanFoot** [comment]
Looking over this looks okay, don't have time to test rn. Nice work!

**Comment by TornaxO7** [comment]
I added moved the actual event loop into the function to separate it better.

File: leftwm-core/src/event_loop.rs [hunk 5]
```diff
@@ -27,12 +27,18 @@ impl<C: Config, SERVER: DisplayServer> Manager<C, SERVER> {
...
-        let mut command_pipe = get_command_pipe().await?;
+        let state_socket = get_state_socket().await?;
+        let command_pipe = get_command_pipe().await?;
 
         self.call_up_scripts();
+        self.start_event_loop(state_socket, command_pipe).await
+    }
 
-        // Main event loop.
+    async fn start_event_loop(
```
Comment (VuiMuich) [inline leftwm-core/src/event_loop.rs:37]: isn't the naming here just backwards with this commit?

Comment (TornaxO7) [reply leftwm-core/src/event_loop.rs:37]: huh, what do you mean with this?

Comment (VuiMuich) [reply leftwm-core/src/event_loop.rs:37]:
```rust
pub async fn start_event_loop(mut self) -> Result<(), Error> {
        let state_socket = get_state_socket().await?;
        let command_pipe = get_command_pipe().await?;

        self.call_up_scripts();
        self.event_loop(state_socket, command_pipe).await
    }

    async fn event_loop(
```

Comment (TornaxO7) [reply leftwm-core/src/event_loop.rs:37]: ooh, yeah, I can do that quickly

Comment (TornaxO7) [reply leftwm-core/src/event_loop.rs]: done!

**Comment by VuiMuich** [comment]
I did run this branch a couple of hours in `Xephyr` yesterday, without any issues.
If there are no last minute vetos (@AethanFoot, @hertg?) I'm gonna merge later today.

**Pull request merged**